core/src/main/java/org/jruby/internal/runtime/ThreadService.java

```diff
@@ -264,25 +261,24 @@ public void setMainThread(Thread thread, RubyThread rubyThread) {
 
                 rtList.add(entry.getValue());
             }
-
-            return rtList.toArray(new RubyThread[rtList.size()]);
         }
+        return rtList.toArray(new RubyThread[rtList.size()]);
     }
 
-    public synchronized ThreadContext registerNewThread(RubyThread thread) {
+    public ThreadContext registerNewThread(RubyThread thread) {
         ThreadContext context = ThreadContext.newContext(runtime);
-        localContext.set(new SoftReference<ThreadContext>(context));
         context.setThread(thread);
-        ThreadFiber.initRootFiber(context); // may be overwritten by fiber
+        ThreadFiber.initRootFiber(context, thread);
+        localContext.set(new SoftReference<>(context));
         return context;
     }
 
-    public synchronized void associateThread(Object threadOrFuture, RubyThread rubyThread) {
-        rubyThreadMap.put(threadOrFuture, rubyThread);
+    public void associateThread(Object threadOrFuture, RubyThread rubyThread) {
+        rubyThreadMap.put(threadOrFuture, rubyThread); // synchronized
     }
 
-    public synchronized void unregisterThread(RubyThread thread) {
-        rubyThreadMap.remove(Thread.currentThread());
+    public void unregisterThread(RubyThread thread) {
+        rubyThreadMap.remove(Thread.currentThread()); // synchronized
         getCurrentContext().setThread(null);
         localContext.set(null);
     }
```

Comment:
Really these three sync are the only major change here, right? Thinking through it now...

* registerNewThread: there's three state changes: ThreadContext gets its RubyThread, the localContext gets a soft ref to the ThreadContext, and the root fiber for the thread is initialized into the context. All of these are or should be thread-local at this point. This seems ok.
* initRootFiber, called from registerNewThread: all thread-local state. Ok.
* associateThread: only one state change against a synchronized collection. Ok.
* unregisterThread: one state changes against a synchronized collection and then thread-local or thread-specific changes. Ok.

Reply:
yep, that's the gist - the rest is just more explicit passing along stuff we work with ...